File: lib/utils/content_filter.dart
import '../config/constants.dart';

/// Content filter result
class FilterResult {
  final bool isAllowed;
  final String filteredText;
  final String? reason;

  const FilterResult({
    required this.isAllowed,
    required this.filteredText,
    this.reason,
  });
}

/// Content filter for messages — blocks URLs, phone numbers, social handles
class ContentFilter {
  ContentFilter._();

  static final RegExp _urlRegex = RegExp(
    AppConstants.urlPattern,
    caseSensitive: false,
  );

  static final RegExp _phoneRegex = RegExp(
    AppConstants.phonePattern,
  );

  /// Filter a message for prohibited content
  static FilterResult filterMessage(String text) {
    final trimmed = text.trim();

    if (trimmed.isEmpty) {
      return const FilterResult(
        isAllowed: false,
        filteredText: '',
        reason: 'Message cannot be empty',
      );
    }

    // Check for URLs
    if (_urlRegex.hasMatch(trimmed)) {
      return FilterResult(
        isAllowed: false,
        filteredText: trimmed,
        reason: 'Links are not allowed in messages',
      );
    }

    // Check for phone numbers (sequences of 7+ digits)
    final digitsOnly = trimmed.replaceAll(RegExp(r'[^\d]'), '');
    if (digitsOnly.length >= 7) {
      return FilterResult(
        isAllowed: false,
        filteredText: trimmed,
        reason: 'Phone numbers are not allowed in messages',
      );
    }

    // Check for social media keywords
    final lowerText = trimmed.toLowerCase();
    for (final keyword in AppConstants.blockedSocialKeywords) {
      if (lowerText.contains(keyword)) {
        return FilterResult(
          isAllowed: false,
          filteredText: trimmed,
          reason: 'Social media handles are not allowed',
        );
      }
    }

    return FilterResult(
      isAllowed: true,
      filteredText: trimmed,
    );
  }

  /// Validate intent text (same rules but less strict on phone numbers)
  static FilterResult filterIntentText(String text) {
    final trimmed = text.trim();

    if (trimmed.isEmpty) {
      return const FilterResult(
        isAllowed: false,
        filteredText: '',
        reason: 'Intent text cannot be empty',
      );
    }

    if (trimmed.length > AppConstants.maxIntentLength) {
      return FilterResult(
        isAllowed: false,
        filteredText: trimmed,
        reason: 'Intent must be ${AppConstants.maxIntentLength} characters or less',
      );
    }

    // Check for URLs
    if (_urlRegex.hasMatch(trimmed)) {
      return FilterResult(
        isAllowed: false,
        filteredText: trimmed,
        reason: 'Links are not allowed',
      );
    }

    // Check for social keywords
    final lowerText = trimmed.toLowerCase();
    for (final keyword in AppConstants.blockedSocialKeywords) {
      if (lowerText.contains(keyword)) {
        return FilterResult(
          isAllowed: false,
          filteredText: trimmed,
          reason: 'Social media references are not allowed',
        );
      }
    }

    return FilterResult(
      isAllowed: true,
      filteredText: trimmed,
    );
  }
}

File: firebase/functions/src/contentFilter.ts
export const BLOCKED_SOCIAL_KEYWORDS = [
  'insta',
  'instagram',
  'snap',
  'snapchat',
  'twitter',
  'tiktok',
  'facebook',
  'whatsapp',
  'telegram',
  'signal',
  'discord',
  'linkedin',
  'youtube',
  'onlyfans',
];

export const URL_PATTERN = /(https?:\/\/[^\s]+)|(www\.[^\s]+)|([a-zA-Z0-9-]+\.(com|org|net|io|co|me|app|dev|xyz)[^\s]*)/i;
export const PHONE_PATTERN = /(\+?\d{1,4}[\s.-]?)?\(?\d{1,4}\)?[\s.-]?\d{1,4}[\s.-]?\d{1,9}/;

export interface FilterResult {
  isAllowed: boolean;
  filteredText: string;
  reason?: string;
}

export function filterMessage(text: string): FilterResult {
  const trimmed = text.trim();

  if (trimmed.length === 0) {
    return {
      isAllowed: false,
      filteredText: '',
      reason: 'Message cannot be empty',
    };
  }

  // Check for URLs
  if (URL_PATTERN.test(trimmed)) {
    return {
      isAllowed: false,
      filteredText: trimmed,
      reason: 'Links are not allowed in messages',
    };
  }

  // Check for phone numbers
  const digitsOnly = trimmed.replace(/[^\d]/g, '');
  if (digitsOnly.length >= 7) {
    return {
      isAllowed: false,
      filteredText: trimmed,
      reason: 'Phone numbers are not allowed in messages',
    };
  }

  // Check for social media keywords
  const lowerText = trimmed.toLowerCase();
  for (const keyword of BLOCKED_SOCIAL_KEYWORDS) {
    if (lowerText.includes(keyword)) {
      return {
        isAllowed: false,
        filteredText: trimmed,
        reason: 'Social media handles are not allowed',
      };
    }
  }

  return {
    isAllowed: true,
    filteredText: trimmed,
  };
}
